### ISSUE TITLE
Map namespace name to connectionstring

### ISSUE THREAD

**Issue by yvesgoeleven** [Particular/NServiceBus.AzureServiceBus#93]
In order to not leak security sensitive information in connectionstrings out, we need a map of logical level namespace name to physical level namespace connectionstring, so that we only need to pass the logical name around.

This will fix the following:

- [x] Ensure ReplyTo does not include shared access keys https://github.com/Particular/NServiceBus.AzureServiceBus/issues/16
- [x] Ensure there are no SharedAccessKey in Logs https://github.com/Particular/NServiceBus.AzureServiceBus/issues/17

**Comment by SeanFeldman** [comment]
@melkio @yvesgoeleven would it help if we flesh out the api (pseudo code) to see how it will work.

**Comment by Scooletz** [comment]
Similar work should be done in ASQ https://github.com/Particular/NServiceBus.AzureStorageQueues/issues/21
It might be obvious, but I'd add that the backward compatibility is required, so we should accept the addresses with the connection string as well, but sending only with logical.

**Comment by melkio** [comment]
I'm missing something for sure but I'd approach the issue in this way:

- changing configuration API to register namespaces. Something like this:

```csharp
var extensions = new TransportExtensions<AzureServiceBusTransport>(settings);
extensions  
    .UseDefaultTopology()
    .Addressing()
    .NamespacePartitioning()
    .AddNamespace("name", "connection string");
```

- @yvesgoeleven: suggested to add a new property to NamespaceInfo to store namespace name and easyly lookup when needed

```csharp
public class NamespaceInfo
{
    public NamespaceInfo(string name, string connectionString, NamespaceMode mode = NamespaceMode.Active)
    {
        Name = name;
        ConnectionString = connectionString;
        Mode = mode;
    }

    public string Name { get; set; }
    public string ConnectionString { get; set; }
    public NamespaceMode Mode { get; set; }

    ....
}
```

- another option could be to stay on the current NamespaceInfo implementation and introduce a new internal component into transport infrastracture, that is able to translate from logical to physical when needed. Something like:

```csharp
public interface ICanTranslateFromLogicalToPhysicalNamespace
{
    string Translate(LogicalNamespace name)
}
```

**Comment by SeanFeldman** [comment]
> It might be obvious, but I'd add that the backward compatibility is required, so we should accept the addresses with the connection string as well, but sending only with logical.

It depends. When sending to older endpoints, those need to receive the connection strings.

**Comment by SeanFeldman** [comment]
@melkio :+1:  for updated `NamespaceInfo`

**Comment by Scooletz** [comment]
@SeanFeldman You're right.

**Comment by yvesgoeleven** [comment]
Indeed, backward compat is required, we'll need to add a config uption as well to turn this capability on/off

**Comment by SeanFeldman** [comment]
A few assumptions:

* The transport should *never* log raw connection string.
* Never pass raw connection strings to the core (it logs things as-is w/o inspecting).
* Pass `NamespaceInfo` within the transport to have access to both, the _raw_ connection string and it's _alias_ .

Raw connection strings are used on the way in and out (passed in headers, ex: `DefaultBatchedOperationsToBrokeredMessagesConverter.SetReplyToAddress()`). 

1. Default behaviour should to be to send the raw connection strings to allow backwards compatibility. 
1. Supply configuration option to override the default and send/receive connection string _aliases_.

@melkio some additional info on this issue.

**Comment by melkio** [comment]
Closed by #125
